### PR TITLE
feat: generalize constructor generation further

### PR DESCRIPTION
**Pull request by Equilibris**
We need a more general notion of constructor generation. The features we need are:

- The ability to change the output type to the desired type
- The ability to control what a recursive type should be able to be replaces with (for `DeepThunk` generation)
- The ability to add required binders used in for-example explicit types (in the case of `DeepThunk` this is the recall parameter)

In this PR we add these features while keeping the behaviour for the specialized case unchanged. This PR is **non-functional** in place of current code

### CODE REVIEW

File: Qpf/Macro/Data/Constructors.lean [hunk 1]
```diff
@@ -36,8 +41,12 @@ def mkConstructorsWithNameAndType (view : DataView) (shape : Name)
       else
         `(fun $args:ident* => $pointConstructor ($shapeCtor $args:ident*))
 
-    let type : Term := TSyntax.mk <|
-      (ctor.type?.map fun type => Replace.replaceAllStx view.getExpectedType ty type).getD ty
+    let recType := view.getExpectedType
+    let forms := RecursionForm.extract ctor recType
+
+    let x := forms.map $ RecursionForm.replaceRec view.getExpectedType argTy
```
Comment (AtticusKuhn) [inline Qpf/Macro/Data/Constructors.lean:47]: I would prefer a more descriptive name than "x".

File: Qpf/Macro/Data/Constructors.lean [hunk 2]
```diff
@@ -62,6 +71,6 @@ def mkConstructors (view : DataView) (shape : Name) : CommandElabM Unit := do
   let explicit ← view.getExplicitExpectedType
   let nameGen := (·.declName.replacePrefix (←getCurrNamespace) .anonymous)
 
-  mkConstructorsWithNameAndType view shape nameGen explicit
+  mkConstructorsWithNameAndType view shape nameGen explicit explicit #[]
```
Comment (AtticusKuhn) [inline Qpf/Macro/Data/Constructors.lean:74]:
```suggestion
  mkConstructorsWithNameAndType view shape nameGen explicit explicit (binders := #[])
```

In functions with many arguments, I think it is sometimes confusing to track all the arguments by position. Thus, I think it's stylistically bettter to pass by named parameters for the arguements.

**Review by AtticusKuhn — APPROVED** [review]
Looks good to me.

**Comment by alexkeizer** [comment]
> This PR is non-functional in place of current code

I guess what you mean here is that you generalize a bunch of things, but don't use the extra general capabilities, leaving the existing behaviour unchanged? I would reserve the claim "non-functional change" for PRs that really only move around definitions, or change the way somethings is phrased to an alternative, but 100% equivalent statement.

Here, I would instead flag it as: we don't yet use the more general capabilities (but require them for a future PR), so the behaviour of existing code is unchanged.

File: Qpf/Macro/Data/Constructors.lean [hunk 3]
```diff
@@ -17,9 +18,13 @@ partial def countConstructorArgs : Syntax → Nat
...
   that are generic across two input types shape and name.
   Additionally we allow the user to control how names are generated.
+  The binders param allows for adding constant binders to an expression.
 -/
-def mkConstructorsWithNameAndType (view : DataView) (shape : Name)
-    (nameGen : CtorView → Name) (ty : Term) : CommandElabM Unit := do
+def mkConstructorsWithNameAndType
+    (view : DataView) (shape : Name)
+    (nameGen : CtorView → Name) (argTy retTy : Term)
+    (binders : TSyntaxArray ``Parser.Term.bracketedBinder)
```
Comment (alexkeizer) [inline Qpf/Macro/Data/Constructors.lean:26]:
```suggestion
    (binders : TSyntaxArray ``Parser.Term.bracketedBinder := #[])
```
This seems like a good place to use a default value (but feel free to ignore if you disagree)

**Review by alexkeizer — APPROVED** [review]
I agree with Atticus' comment about having a more descriptive name, and would like you to phrase the docstring a bit more clearly.

Otherwise, LGTM

**Pull request merged**